gnocchi/amqp1d.py

```diff
@@ -0,0 +1,251 @@
+# Copyright (c) 2015 eNovance
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import print_function
+import itertools
+import ujson
+import uuid
+
+import daiquiri
+from oslo_config import cfg
+
+import six
+
+from gnocchi import incoming
+from gnocchi import indexer
+from gnocchi import service
+
+from gnocchi import utils
+
+from proton.handlers import MessagingHandler
+from proton.reactor import Container
+
+LOG = daiquiri.getLogger(__name__)
+
+
+class CollectdStats(object):
+    def __init__(self, conf):
+        self.conf = conf
+        self.incoming = incoming.get_driver(self.conf)
+        self.indexer = indexer.get_driver(self.conf)
+        self.resource_type = None
+
+        self._ensure_resource_type_exists()
+        self.gauges = {}
+        self.counters = {}
+        self.absolute = {}
+
+    def reset(self):
+        self.gauges.clear()
+        self.counters.clear()
+        self.absolute.clear()
+
+    def on_timer_task(self, event):
+        event.container.schedule(self.conf.amqp1d.flush_delay, self)
+        self.flush()
+
+    @staticmethod
+    def serialize_identifier(index, value):
+        """Based of FORMAT_VL from collectd/src/daemon/common.h.
+
+        The biggest difference is that we don't prepend the host and append the
+        index of the value, and don't use slash.
+
+        """
+        return (value["plugin"] + ("-" + value["plugin_instance"]
+                                   if value["plugin_instance"] else "")
+                + "@"
+                + value["type"] + ("-" + value["type_instance"]
+                                   if value["type_instance"] else "")
+                + "-" + str(index))
+
+    def treat_metric(self, host, metric_name, metric_type,
+                     value):
+        """Collectd.
+
+        Statistics in collectd consist of a value list. A value list includes:
+        Values, can be one of:
+        Derive: used for values where a change in the value since it's last
+        been read is of interest. Can be used to calculate and store a rate.
+        Counter: similar to derive values, but take the possibility of a
+        counter wrap around into consideration.
+        Gauge: used for values that are stored as is.
+        Absolute: used for counters that are reset after reading.
+
+        """
+
+        if metric_type == "absolute":
+            if host not in self.absolute:
+                self.absolute[host] = {}
```

Comment:
you can use a defaultdict in that case:

`self.absolute = collections.defaultdict(dict)`